### PR TITLE
common: add iterator-based string splitter

### PR DESCRIPTION
**Pull request by cbodley**
ranged-for loop example:
```c++
  for (std::string_view s : split(input)) {
    ...
```
container initialization example:
```c++
  auto parts = split(input);

  std::vector<std::string> strings;
  strings.assign(parts.begin(), parts.end());
```

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/split.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ */
+
+#pragma once
+
+#include <string_view>
+
+namespace ceph {
+
+// a forward iterator over the parts of a split string
+class spliterator {
```
Comment (tchaikov) [inline src/common/split.h]: how about moving these helpers into `str_list.h` , so we can start deprecating the `for_each_substr()` and friends?

Comment (adamemerson) [reply src/common/split.h]: I'd suggest keeping it out of str_list (the name doesn't really fit) and eventually planning on removing that whole file.

Comment (cbodley) [reply src/common/split.h]: is there a reason it needs to be in str_list.h before we can deprecate those?

Comment (tchaikov) [reply src/common/split.h]: not "needs to be", just thought it'd be easier to redirect new users to these `split*` utilies.

File: src/test/common/test_split.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  return std::equal(lhs.begin(), lhs.end(), rhs.begin(), rhs.end());
+}
+bool operator==(const string_list& lhs, const split& rhs) {
+  return std::equal(lhs.begin(), lhs.end(), rhs.begin(), rhs.end());
+}
+
+TEST(split, split)
+{
+  EXPECT_EQ(string_list({}), split(""));
+  EXPECT_EQ(string_list({}), split(","));
```
Comment (batrick) [inline src/test/common/test_split.cc]: String splitting in most languages would return two empty strings here.

File: src/test/common/test_split.cc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  EXPECT_EQ(string_list({}), split(",;"));
+
+  EXPECT_EQ(string_list({"a"}), split("a,;"));
```
Comment (batrick) [inline src/test/common/test_split.cc]: Continuation of previous comment, I would expect:

`{"a", "", ""}`

File: src/test/common/test_split.cc [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+{
+  const auto parts = split("a,b");
+  auto i = parts.begin();
+  ASSERT_NE(i, parts.end());
+  ASSERT_EQ(2, std::distance(i, parts.end()));
+
+  // copy the iterator to test LegacyForwardIterator's multipass guarantee
+  auto j = i;
+  ASSERT_EQ(i, j);
+  ASSERT_NE(j, parts.end());
```
Comment (batrick) [inline src/test/common/test_split.cc]: These tests should really be broken into separate test cases. It's easier to follow that way and avoid dependencies between tests.

File: src/test/common/test_split.cc (hunk 2 above)
Comment (adamemerson) [reply src/test/common/test_split.cc]: The interface we're replacing merges runs of delimiters.

File: src/test/common/test_split.cc (hunk 4 above)
Comment (cbodley) [reply src/test/common/test_split.cc]: thanks, i split the iterator test into 6 more specific ones

**Comment by cbodley** [comment]
thanks for reviews - merging without qa because it's only used in the unit test

**Pull request merged**